website/blog/2020-02-24-presto-lambda.md

```diff
@@ -0,0 +1,193 @@
+FROM (
+    VALUES
+        (ARRAY[1, 2], 2),
+        (ARRAY[1, 3, 9], 3),
+        (ARRAY[1, 4, 16, 64], 4)
+) AS t(elements, factor)
+
+    elements    | multiplied_elements
+----------------+---------------------
+ [1, 2]         | [2, 4]
+ [1, 3, 9]      | [3, 9, 27]
+ [1, 4, 16, 64] | [4, 16, 64, 256]
+(3 rows)
+```
+
+Lambda capture supported is added in [Presto#7210](https://github.com/prestodb/presto/pull/7210). It rewrites the captured lambda into non-capture lambda via [partial function application](https://en.wikipedia.org/wiki/Partial_application). A special internal higher order function `BIND` is introduced to partially apply captured arguments to the lambda.
+
+Take the above example, the captured lambda call
+
+```sql
+transform(elements, v -> v * factor)
+```
+
+is rewritten to
+
+```sql
+transform(
+    elements, 
+    BIND(factor, (captured_factor, e) -> e * captured_factor)
+)
+```
+
+The original unary lambda with capture `e -> e * factor` is rewritten into a binary lambda without capture: `(captured_factor, e) -> e * captured_factor`. The `BIND` call takes `factor` and this binary lambda as input, returns the partially applied function that multiplies the input by `captured_factor` (Note the `captured_factor` will be different for each row!). This partially applied function is an unary function and is provided as the second parameter to `transform` call.
+
+## Lambda Execution
+
+In this section we are going to discuss how lambda is executed during runtime. The original implementation uses `MethodHandle` object to represent lambda on stack. Consider the same example:
+
+```sql
+transform(
+    elements, 
+    BIND(factor, (captured_factor, v) -> v * captured_factor)
+)
+```
+
+Each invocation of `transform` works in the following way: 
+1. Push the Java object represents elements on the stack. When `elements` is `array(integer)`, the corresponding Java stack type is `IntArrayBlock`.
+2. Push the `MethodHandle` object represents captured lambda to the stack, i.e. binding `factor` to `(captured_factor, v) -> v * captured_factor`. To this end: 
+    1. Push the `MethodHandle` object represents `v -> v * captured_factor` on the stack.
+    2. Push `captured_factor` on the stack.
+    3. Invoke `MethodHandle#bindTo` to get a `BoundMethodHandle` represents captured lambda on the top of stack. 
+3. Invoke `transform`. 
+
+Unfortunately, this implementation causes Java to generate a separate customized LambdaForm class for every `MethodHandle#bindTo` call (i.e. per each row). Such excessive runtime class generation quickly fills in metaspace full and causes full GC, see [Presto#7935](https://github.com/prestodb/presto/issues/7935) for reproduction and details. JDK developers also confirms that it is expected to have each `BoundMethodHandle` to be customized independently, and `MethodHandle#bindTo` [is not a good fit for implementing lambda capturing](http://mail.openjdk.java.net/pipermail/mlvm-dev/2017-May/006755.html).
```

Comment:
JDK developers have confirmed that each `BoundMethodHandle` should be customized independently, and